### ISSUE TITLE
Enhancement: [`prefer-optional-chain`] handle cases where the first logical(s) are unrelated to the proceeding logicals

### ISSUE THREAD

**Issue by B2o5T** [typescript-eslint/typescript-eslint#6332]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-optional-chain

### Description

here is totally safe to refactor as I mention in `Pass`

### Fail

```typescript
if (!baz || !foo || !foo.bar) {
  
}
```

### Pass

```typescript
if (!baz || !foo?.bar) {
  
}
```

### Additional Info

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=JYMwBAFAhARghgLzAH2WKID2mVo9gOngCcBKMAbwCgwwqBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uki9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Comment by bradzacher** [comment]
The rule does catch the logical OR case, the issue is that the rule does not catch cases with unrelated logicals at the beginning ([playground](https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAK4qjYCWKIw8APvAN7wBGshmMlA5vAL4DcAKEHls8ABQBCXEjr1peAHSsYASkaD48Qb0FA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uki9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)).

Why? Because it's just a really, really, really hard problem to solve using the AST.
To illustrate this, take your case like this
```
a || b || b.c
```

For a human this is really easy to parse and understand, but the AST this produces is this:
```
{
  type: LogicalExpression,
  operator: '||',
  left: b.c
  right: {
    type: LogicalExpression,
    operator: '||',
    left: a,
    right: b,
  }
}
```

It's difficult to properly associate the fact that `b` and `b.c` are positioned sequentially in the logical in the general case. Parentheses and mixed logicals further complicate this by changing the AST shape due to changing precedence.
Combined with the fact the rule needs to also understand the expressions can be of the form `!expr`, `expr`, `expr != null`, `expr !== undefined && expr !== null`, ...

It's just a hard problem to solve in the general case!

**Comment by B2o5T** [comment]
Hey @bradzacher! Many thanks for the detailed answer!

> Why? Because it's just a really, really, really hard problem to solve using the AST.

to be honest is not! and can be solved easily with `flatLogicalExpression` from `eslint-plugin-unicorn`

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/df4295b53eb33f7c34b5bd81c1c3ce50192bcb65/rules/no-useless-length-check.js#L35-L41

I remember 1 year ago I wrote a rule `prefer-array-includes` to do the following and logical expressions are not too scary 😂

Code

    1 | foo === null || foo !== 7 || ![8, 9].includes(foo)

Error 1/1

    > 1 | foo === null || foo !== 7 || ![8, 9].includes(foo)
        |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use \`.includes()\`, rather than repeated conditional logical OR \`||\` operators.

Output

    1 | foo === null || ![7, 8, 9].includes(foo)

**Comment by bradzacher** [comment]
It's funny cos when I was writing my comment I was thinking that the entire rule could probably be refactored to do exactly that - collect the expressions to remove the AST from the equation.

Theres probably a lot of abstractions that could be built into the rule to improve it and reduce the verbosity of the code.

It's grown a lot over time. Again - happy to accept a PR!

**Comment by Josh-Cena** [comment]
Fixed by https://github.com/typescript-eslint/typescript-eslint/pull/6397
